synthtool/gcp/gapic_microgenerator.py

```diff
@@ -0,0 +1,199 @@
+        """
+        Generates the Python Library files using artman/GAPIC
+        returns a `Path` object
+        library: path to library. 'google/cloud/speech'
+        version: version of lib. 'v1'
+        """
+        return self._generate_code(service, version, "python", **kwargs)
+
+    def go_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "go", **kwargs)
+
+    def kotlin_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "kotlin", **kwargs)
+
+    def _generate_code(
+        self,
+        service: str,
+        version: str,
+        language: str,
+        *,
+        private: bool = False,
+        proto_path: str = None,
+        output_dir: str = None,
+        generator_version: str = "latest",
+        generator_args: Mapping[str, str] = None,
+    ):
+        # Determine which googleapis repo to use
+        if not private:
+            googleapis = self._clone_googleapis()
+        else:
+            googleapis = self._clone_googleapis_private()
+
+        # Sanity check: We should have a googleapis repo; if we do not,
+        # something went wrong, and we should abort.
+        if googleapis is None:
+            raise RuntimeError(
+                f"Unable to generate {service}, the googleapis repository"
+                "is unavailable."
+            )
+
+        # Pull the code generator for the requested language.
+        # If a code generator version was specified, honor that.
+        log.debug("Pulling Docker image: gapic-generator-{language}:{generator_version}")
+        shell.run([
+            "docker",
+            "pull",
+            f"gcr.io/gapic-images/gapic-generator-{language}:{generator_version}",
+        ], hide_output=False)
+
+        # Determine where the protos we are generating actually live.
+        # We can sometimes (but not always) determine this from the service
+        # and version; in other cases, the user must provide it outright.
+        if proto_path:
+            proto_path = Path(proto_path)
+            if proto_path.is_absolute():
+                proto_path = proto_path.relative_to("/")
+        else:
+            proto_path = Path("google/cloud") / service / version
+
+        # Sanity check: Do we have protos where we think we should?
+        if not (googleapis / proto_path).exists():
+            raise FileNotFoundError(
+                f"Unable to find directory for protos: {(googleapis / proto_path)}."
+            )
+        if not tuple((googleapis / proto_path).glob("*.proto")):
+            raise FileNotFoundError(
+                f"Directory {(googleapis / proto_path)} exists, but no protos found."
+            )
+
+        # Ensure the desired output directory exists.
+        # If none was provided, create a temporary directory.
+        if not output_dir:
+            output_dir = tempfile.mkdtemp()
```

Comment:
@crwilcox I think it is pretty likely that this is the wrong thing to do, but it is not obvious to me what a _better_ option is.

Here are a couple potential possibilities:

  * Return a `TemporaryDirectory` object _instead of_ a `Path`.
    * _Pros:_ Easy for `synth.py` files; they just use the context manager interface to copy out of the temporary directory and cleanup magically happens.
    * _Cons:_ Violates / changes the interface observed by `GAPICGenerator`.
  * Require an output directory.
    * _Pros:_ Unambiguous. If you create a temporary directory to handle this, you clean it up.
    * _Cons:_ Slightly more code than the option above (~1-2 more lines in each `synth.py` file). Slightly violates the interface by requiring `output_dir`.
  * What is here now.
    * _Pros:_ Holds to the existing interface.
    * _Cons:_ Very easy to leave temporary directories lying around.
  * Try to return a frankenstein object that subclasses both `Path` and `TemporaryDirectory`. (In this case, I would remove `output_dir` entirely and _only_ write to temp directories.)
    * _Pros:_ Holds to the existing interface.
    * _Cons:_ Seems dirty. Edge cases intuitively seem likely.

Reply:
I am OK with the approach here. Synthtool already leaves behind generated libraries in `~/.cache/synthtool/googleapis/artman-genfiles/`.

In fact, we have a note in the README that recommends taking a look at that directory when you're doing a generation.

> SynthTool will run Artman which will create generated code that can be found at `~/.cache/synthtool/googleapis<-private>/artman_genfiles`. This is useful for figuring out what it is you need to copy for your specific library.

Would it be reasonable to have the microgenerator default to a similar location for predictability?

*For reference:*
Artman's output_dir is `root_dir / "artman-genfiles"`
https://github.com/googleapis/synthtool/blob/6f7311c3a6fc555f81fb0bcb08a4596e11e98608/synthtool/gcp/artman.py#L34-L43

`root_dir` is the path for `googleapis`.
https://github.com/googleapis/synthtool/blob/6f7311c3a6fc555f81fb0bcb08a4596e11e98608/synthtool/gcp/gapic_generator.py#L108-L112

And googleapis is cloned to a cache dir, by default.
https://github.com/googleapis/synthtool/blob/6f7311c3a6fc555f81fb0bcb08a4596e11e98608/synthtool/sources/git.py#L47-L55
https://github.com/googleapis/synthtool/blob/6f7311c3a6fc555f81fb0bcb08a4596e11e98608/synthtool/cache.py#L19

Reply:
Thanks @busunkim96 for looking at this!

The new generators are protoc plugins, which require a specified directory to output to (that is the fundamental protoc plugin interface), so I can not have the generator itself default to any location.

However, I _can_ have synthtool pick a default location (e.g. `python-gapic-genfiles`) and behave this way if that is your preference. That seems odd to me, but you are the end user, and I defer to your preference. :)

Should I make that change? Happy to do so.

Reply:
@lukesneeringer Yes please! I think it would be the least surprising for the microgenerator to also put files in the synthtool cache dir. Thank-you. :smile: